File: java-ds-and-algorithms/src/main/java/sort/QuickSort.java
package sort;

/**
 * Usage: 快速排序 <br/>
 * Date: 2023/5/17 10:11 <br/>
 *
 * @author <a href="mailto:hanzhang2566@foxmail.com">hanzhang</a>
 */
public class QuickSort {
    private final int[] nums;

    public QuickSort(int[] nums) {
        this.nums = nums;
    }

    public void sort(int head, int tail) {
        if (head >= tail) {
            return;
        }
        int pivotIndex = partition(head, tail);
        sort(head, pivotIndex - 1);
        sort(pivotIndex + 1, tail);
    }

    public int select(int head, int tail, int targetIndex) {
        if (head >= tail) {
            return -1;
        }
        int pivotIndex = partition(head, tail);
        if (pivotIndex > targetIndex) {
            return select(head, pivotIndex - 1, targetIndex);
        } else if (pivotIndex < targetIndex) {
            return select(pivotIndex + 1, tail, targetIndex);
        } else {
            return pivotIndex;
        }
    }

    private int partition(int head, int tail) {
        int left = head;
        int right = tail;

        while (left < right) {
            while (left < right && nums[left] < nums[head]) {
                left++;
            }
            while (left < right && nums[right] > nums[head]) {
                right--;
            }
            swap(nums, left, right);
        }
        swap(nums, left, head);
        return left;
    }

    public void swap(int[] nums, int left, int right) {
        int temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
    }
}

File: go-ds-and-algorithms/sort/quick_sort.go
package sort

func QuickSort(nums []int, head, tail int) {
	if head >= tail {
		return
	}
	pivotPos := partition(nums, head, tail)
	QuickSort(nums, head, pivotPos-1)
	QuickSort(nums, pivotPos+1, tail)
}

// 将切片进行一次快速排序，返回分隔索引
func partition(nums []int, head, tail int) int {
	left := head
	right := tail
	for left < right {
		for left < right && nums[left] < nums[head] {
			left++
		}
		for left < right && nums[right] > nums[head] {
			right--
		}
		nums[left], nums[right] = nums[right], nums[left]
	}
	nums[left], nums[head] = nums[head], nums[left]
	return left
}

// 二分查找和快排的结合
func quickSelect(nums []int, head, tail, targetIndex int) int {
	if head >= tail {
		return -1
	}

	index := partition(nums, head, tail)
	if targetIndex < index {
		return quickSelect(nums, head, index-1, targetIndex)
	} else if targetIndex > index {
		return quickSelect(nums, index+1, tail, targetIndex)
	} else {
		return index
	}
}
